### ISSUE TITLE
Parallel Data Writing in HDF5 Output

### ISSUE THREAD

**Issue by LTMeyer** [pdebench/PDEBench#17]
Hi,

First let me thank you for the great initiative to create dataset and benchmark for machine learning applied to PDEs.

I tried to generate data using your script : 
`python3 data_gen/gen_diff_react.py`
I got the following error:
> BlockingIOError: [Errno 11] Unable to open file (unable to lock file, errno = 11, error message = 'Resource temporarily unavailable')

I assume this error arises because at least two processes of the pool are trying to write in the HDF5 output at the same time. HDF5 actually supports parallel writing (c.f. [official documentation](https://docs.h5py.org/en/stable/mpi.html#parallel-hdf5)). This supports however relies on mpi. Hence, I'm not sure the `multirpocessing.Pool` allows the parallel writing of the simulation data in the output file. Let me know if such explanation is erroneous. 

Which configuration do you use to be able to successfully generate and write the data in parallel? Which `h5py` installation do you use (mpi supported or standard one)?

A quick fix to the problem occurring is to retry the opening of the file until the lock is available: 
```python
busy = True
while busy:
    try:
        data_f = h5py.File(utils.expand_path(config.output_path), "a")
    except IOError: 
        time.sleep(1)
        print(f"Busy {busy}.")
        continue
    else:
        ...  # Writing of the dataset
        data_f.close()

```

Nonetheless, proper support of parallel HDF5 writing would be appreciated.

**Comment by leiterrl** [comment]
Thanks for raising this issue. This has probably slipped through the cracks since we used to generate seperate files for each simulation run and implemented this last minute. I suspect proper parallel HDF5 writing support is a little out of scope for this project as it adds too many complicated installation steps for our users.

Meanwhile, I tested your solution of just waiting until the file lock is released but still ran into different filesystem related exceptions after generating >1000 data points. I think we will probably revert back to our previous approach of generating separate files soon.

**Comment by LTMeyer** [comment]
Thank you for the reply. 

What were the filesystem exceptions you obtained? I encountered issues while generating many data points, which was related to GPU overflow memory rather than filesystem issue. It occurred even when multiple GPUs were available.

In any case, it would be nice if the data generation was unified and working. Could you please let me know in the current issue when the previous approach of generating files is restored?

**Comment by leiterrl** [comment]
I added the workaround as you proposed it for now. Seems like my issues were just caused by working on network storage. I can reference this issue once we restructure the code in the future.
